rust/worker/src/execution/operators/filter.rs

```diff
@@ -3,117 +3,92 @@ use std::{
     ops::{BitAnd, BitOr, Bound},
 };
 
-use crate::{
-    execution::operator::Operator,
-    segment::{
-        metadata_segment::{MetadataSegmentError, MetadataSegmentReader},
-        record_segment::{RecordSegmentReader, RecordSegmentReaderCreationError},
-        LogMaterializer, LogMaterializerError, MaterializedLogRecord,
-    },
-};
-use chroma_blockstore::provider::BlockfileProvider;
 use chroma_error::{ChromaError, ErrorCodes};
 use chroma_index::metadata::types::MetadataIndexError;
 use chroma_types::{
     BooleanOperator, Chunk, DirectDocumentComparison, DirectWhereComparison, DocumentOperator,
-    LogRecord, MaterializedLogOperation, MetadataSetValue, MetadataValue, PrimitiveOperator,
-    Segment, SetOperator, SignedRoaringBitmap, Where, WhereChildren, WhereComparison,
+    MaterializedLogOperation, MetadataSetValue, MetadataValue, PrimitiveOperator, SetOperator,
+    SignedRoaringBitmap, Where, WhereChildren, WhereComparison,
 };
 use roaring::RoaringBitmap;
 use thiserror::Error;
 use tonic::async_trait;
 use tracing::{trace, Instrument, Span};
 
-/// # Description
-/// The `MetadataFilteringOperator` should produce the offset ids of the matching documents.
-///
-/// # Input
-/// - `blockfile_provider` / `record_segment` / `metadata_segment`: handles to the underlying data.
-/// - `log_record`: the chunk of log that is not yet compacted, representing the latest updates.
-/// - `query_ids`: user provided ids, which must be a superset of returned documents.
-/// - `where_clause`: a boolean predicate on the metadata and the content of the document.
-///
-/// # Output
-/// - `log_record`: the same `log_record` from the input.
-/// - `log_oids`: the offset ids to include or exclude from the log.
-/// - `compact_oids`: the offset ids to include or exclude from the record segment.
-///
-/// # Note
-/// - The `MetadataProvider` enum can be viewed as an universal interface for the metadata and document index.
-/// - In the output, `log_mask` should be a subset of `offset_ids`
+use crate::{
+    execution::operator::Operator,
+    segment::{
+        metadata_segment::MetadataSegmentReader, LogMaterializer, LogMaterializerError,
+        MaterializedLogRecord,
+    },
+};
 
-#[derive(Debug)]
-pub struct FilterOperator {}
+use super::{
+    fetch_log::FetchLogOutput,
+    fetch_segment::{FetchSegmentError, FetchSegmentOutput},
+};
 
-impl FilterOperator {
-    pub fn new() -> Box<Self> {
-        Box::new(FilterOperator {})
-    }
+#[derive(Clone, Debug)]
+pub struct FilterOperator {
+    pub query_ids: Option<Vec<String>>,
+    pub where_clause: Option<Where>,
 }
 
-#[derive(Debug)]
-pub struct FilterInput {
-    blockfile_provider: BlockfileProvider,
-    record_segment: Segment,
-    metadata_segment: Segment,
-    log_record: Chunk<LogRecord>,
-    query_ids: Option<Vec<String>>,
-    where_clause: Option<Where>,
+#[derive(Clone, Debug, Default)]
+pub struct PreFilterState {
```

Comment:
what purpose does this serve?

Reply:
This keeps a partial result from the `FetchLogOperator` and `FetchSegmentOperator`. Only when we have results from both can we proceed to `FilterOperator`

Reply:
Update: I'll simply keep these fields inside the orchestrator.